File: src/java/magma/api/collect/HeadedQuery.ts
import { Tuple2 } from "../../../magma/api/Tuple2";
import { Option } from "../../../magma/api/option/Option";
import { Main } from "../../../magma/app/Main";
import { BiFunction } from "../../../java/util/function/BiFunction";
import { Function } from "../../../java/util/function/Function";
import { Predicate } from "../../../java/util/function/Predicate";
export class HeadedQuery<T> implements Query<T> {
	head: Head<T>;
	constructor (head: Head<T>) {
		this.head = head;
	}
	next(): Option<T> {
		return this.head.next();
	}
	collect<C>(collector: Main.Collector<T, C>): C {
		return this.foldWithInitial(collector.createInitial(), collector.fold);
	}
	map<R>(mapper: (arg0 : T) => R): Query<R> {
		return new HeadedQuery<R>(new Main.MapHead<T, R>(this.head, mapper));
	}
	foldWithInitial<R>(initial: R, folder: (arg0 : R, arg1 : T) => R): R {
		let result: R = initial;
		while (true){
			let finalResult: R = result;
			let maybeNext: Tuple2<Boolean, R> = this.head.next().map((inner: T) => folder(finalResult, inner)).toTuple(finalResult);
			if (maybeNext.left()){
				result = maybeNext.right();
			}
			else {
				return result;
			}
		}
	}
	foldWithMapper<R>(next: (arg0 : T) => R, folder: (arg0 : R, arg1 : T) => R): Option<R> {
		return this.head.next().map(next).map((maybeNext: R) => {
			return this.foldWithInitial(maybeNext, folder);
		});
	}
	flatMap<R>(mapper: (arg0 : T) => Query<R>): Query<R> {
		return this.head.next().map(mapper).map((initial: Query<R>) => new HeadedQuery<R>(new Main.FlatMapHead<T, R>(this.head, initial, mapper))).orElseGet(() => new HeadedQuery<R>(new Main.EmptyHead<R>()));
	}
	allMatch(predicate: (arg0 : T) => boolean): boolean {
		return this.foldWithInitial(true, (maybeAllTrue: Boolean, element: T) => maybeAllTrue && predicate(element));
	}
	filter(predicate: (arg0 : T) => boolean): Query<T> {
		return this.flatMap((element: T) => {
			if (predicate(element)){
				return new HeadedQuery<T>(new Main.SingleHead<T>(element));
			}
			else {
				return new HeadedQuery<T>(new Main.EmptyHead<T>());
			}
		});
	}
}

File: src/windows/magma/api/collect/head/HeadedQuery.c
#include "./HeadedQuery.h"
export class HeadedQuery<T> implements Query<T> {
	Head<T> head;
	constructor (Head<T> head) {
		this.head = head;
	}
}

Option<T> next() {
	return this.head.next();
}
C collect(Collector<T, C> collector) {
	return this.foldWithInitial(collector.createInitial(), (C current, T element) => collector.fold(current, element));
}
Query<R> map((arg0 : T) => R mapper) {
	return new HeadedQuery<R>(new MapHead<T, R>(this.head, mapper));
}
R foldWithInitial(R initial, (arg0 : R, arg1 : T) => R folder) {
	R result = initial;
	while (true) {
		R finalResult = result;
		Tuple2<Bool, R> maybeNext = this.head.next().map((T inner) => folder(finalResult, inner)).toTuple(finalResult);
		if (maybeNext.left()) {
			result = maybeNext.right();
		}
		else {
			return result;
		}
	}
}
Option<R> foldWithMapper((arg0 : T) => R next, (arg0 : R, arg1 : T) => R folder) {
	return this.head.next().map(next).map((R maybeNext) => this.foldWithInitial(maybeNext, folder));
}
Query<R> flatMap((arg0 : T) => Query<R> mapper) {
	return this.head.next().map(mapper).map((Query<R> initial) => new HeadedQuery<R>(new FlatMapHead<T, R>(this.head, initial, mapper))).orElseGet(() => new HeadedQuery<R>(new EmptyHead<R>()));
}
Bool allMatch((arg0 : T) => boolean predicate) {
	return this.foldWithInitial(true, (Bool maybeAllTrue, T element) => maybeAllTrue && predicate(element));
}
Bool anyMatch((arg0 : T) => boolean predicate) {
	return this.foldWithInitial(false, (Bool aBoolean, T t) => aBoolean || predicate(t));
}
Query<Tuple2<T, R>> zip(Query<R> other) {
	return new HeadedQuery<Tuple2<T, R>>(new ZipHead<T, R>(this.head, other));
}
Query<T> filter((arg0 : T) => boolean predicate) {
	return this.flatMap((T element) => {
		if (predicate(element)) {
			return new HeadedQuery<T>(new SingleHead<T>(element));
		}
		else {
			return new HeadedQuery<T>(new EmptyHead<T>());
		}
	});
}